core/src/test/java/org/elasticsearch/cluster/action/shard/ShardFailedClusterStateTaskExecutorTests.java

```diff
@@ -0,0 +1,204 @@
+        RoutingNodes routingNodes = stateAfterReroute.getRoutingNodes();
+        RoutingTable afterStart =
+            allocationService.applyStartedShards(stateAfterReroute, routingNodes.shardsWithState(ShardRoutingState.INITIALIZING)).routingTable();
+        return ClusterState.builder(stateAfterReroute).routingTable(afterStart).build();
+    }
+
+    private List<ShardStateAction.ShardRoutingEntry> createExistingShards(ClusterState currentState, String reason) {
+        List<ShardRouting> shards = new ArrayList<>();
+        GroupShardsIterator shardGroups =
+            currentState.routingTable().allAssignedShardsGrouped(new String[] { INDEX }, true);
+        for (ShardIterator shardIt : shardGroups) {
+            for (ShardRouting shard : shardIt.asUnordered()) {
+                shards.add(shard);
+            }
+        }
+        List<ShardRouting> failures = randomSubsetOf(randomIntBetween(1, 1 + shards.size() / 4), shards.toArray(new ShardRouting[0]));
+        String indexUUID = metaData.index(INDEX).getIndexUUID();
+        int numberOfTasks = randomIntBetween(failures.size(), 2 * failures.size());
+        List<ShardRouting> shardsToFail = new ArrayList<>(numberOfTasks);
+        for (int i = 0; i < numberOfTasks; i++) {
+            shardsToFail.add(randomFrom(failures));
+        }
+        return toTasks(shardsToFail, indexUUID, reason);
+    }
+
+    private List<ShardStateAction.ShardRoutingEntry> createNonExistentShards(String reason) {
+        MetaData nonExistentMetaData =
+            MetaData.builder()
+                .put(IndexMetaData.builder("non-existent").settings(settings(Version.CURRENT)).numberOfShards(1).numberOfReplicas(numberOfReplicas))
+                .build();
+        RoutingTable routingTable = RoutingTable.builder().addAsNew(nonExistentMetaData.index("non-existent")).build();
+        String nonExistentIndexUUID = nonExistentMetaData.index("non-existent").getIndexUUID();
+        return toTasks(routingTable.allShards(), nonExistentIndexUUID, reason);
+    }
+
+    private static void assertTasksSuccessful(
+        List<ShardStateAction.ShardRoutingEntry> tasks,
+        ClusterStateTaskExecutor.BatchResult<ShardStateAction.ShardRoutingEntry> result,
+        ClusterState clusterState,
+        boolean clusterStateChanged
+    ) {
+        Map<ShardStateAction.ShardRoutingEntry, Boolean> taskResultMap =
+            tasks.stream().collect(Collectors.toMap(Function.identity(), task -> true));
+        assertTaskResults(taskResultMap, result, clusterState, clusterStateChanged);
+    }
+
+    private static void assertTaskResults(
+        Map<ShardStateAction.ShardRoutingEntry, Boolean> taskResultMap,
+        ClusterStateTaskExecutor.BatchResult<ShardStateAction.ShardRoutingEntry> result,
+        ClusterState clusterState,
+        boolean clusterStateChanged
+    ) {
+        assertEquals(taskResultMap.size(), result.executionResults.size());
```

Comment:
do we want to check that the shards were actually failed? Ideally we would just check that the allocation service was asked to fail them. Mockito maybe helpful? I don't want to hold you back for this though...